### PR TITLE
full CMake port & support

### PR DESCRIPTION
**Pull request by mykola2312**
Dear [xaionaro](https://github.com/xaionaro),

I submitted this PR because your driver needs to be ported to CMake ASAP, so it can be used in more advanced IDEs with CMake support.
I have made some changes to code and moved all config definitions to CMake variables, as well as making this driver fully able to build as library against other projects using [stm32-cmake](https://github.com/ObKo/stm32-cmake) and its template form [stm32-base](https://github.com/DVALight/stm32-base).
Test project using this PR You can see at [stm32-test](https://github.com/DVALight/stm32-test/tree/test-for-pr-cmake).

Thank You for initial EtherShield port from Arduino.

Best regards,

[mykola2312](https://github.com/mykola2312)

### CODE REVIEW

File: README.md
Comment (xaionaro) [inline README.md]: Could you also share a link on an example how to use this library (since the old examples are obsolete and removed)? May be something like https://github.com/DVALight/stm32-test/blob/test-for-pr-cmake/src/main.c ?

File: CMakeLists.txt
Comment (xaionaro) [inline CMakeLists.txt:89]: Out of curiosity, what is the purpose of these lines? :)

**Comment by xaionaro** [comment]
O, wow, this is a big PR. Thanks! This is unfortunate I don't use STM32 anymore and cannot test this and cannot review this properly. Also it seems that this PR is breaking (people using this library won't be able to update without breaking their code).

Having said all that, I think I'll create a separate branch, merge it there, and just switch the default git branch to the new branch. And someday will delete the old branch. In the meanwhile, could you please take a look at the couple of comments I left in the review?

File: CMakeLists.txt
Comment (mykola2312) [reply CMakeLists.txt:89]: its command for converting ELF to intel HEX, because some simulators like Proteus run only hex files. You need that command only on final firmware target

File: README.md
Comment (mykola2312) [reply README.md]: Examples that you mentioned in README are good and works fine with my PR, but unfortunately they are on russian git site that I cant access even with Tor. It would be awesome if you reupload them on github so I can add in PR's README

Comment (xaionaro) [reply README.md]: I see. I guess RKN or something is continuing doing insane stuff. I've just cloned repositories to:
* https://github.com/mephi-ut/dc-thermal-logger
* https://github.com/mephi-ut/STM32_Devel
* https://github.com/mephi-ut/nucleo-f091rc-rs232-enc28j60

Comment (mykola2312) [reply README.md]: Thanks. I will add examples to PR's readme

Comment (mykola2312) [reply README.md]: I have updated readme.

**Review by xaionaro — APPROVED** [review]

**Comment by xaionaro** [comment]
@mykola2312 : Hey! I found some time to test it, and found that the code was hardcoded to STMF4. Sorry, but I had to neutralize those pieces and add comment about this problem back to README.md (as it was in v1). If you feel this is wrong, feel free to create a one more PR :)

**Comment by mykola2312** [comment]
Thats a bummer. Why you moved headers to inc folder?

**Comment by xaionaro** [comment]
Forgot to say one more thing: since you are the only active developer of this library in the whole world right now, it might be reasonable to adding you to collaborators, and may be move to a neutral org (so that it is not coupled with `xaionaro`) :)

About `inc`: I should've asked about `inc`-vs-`src` in the PR before merging it. The reason why I reverted it is just because I haven't found the reason why they were merged, while older projects already have configurations with `inc` (and I just didn't want to break them without reason). But if there is a good reason I'm happy to merge it back :)

**Comment by mykola2312** [comment]
Why people keep following that nonsense from CubeIDE... Anyway, you removed stm32 hal conf header.. which hasld important modules enabled. I suggest you revert changes about inc folder and removing hal conf files - because that change will made it even harder for people to adapt it for their MCU

**Comment by mykola2312** [comment]
Im planning to make next huge PR with refactored code and some changes. I will keep in mind requirements for other MCUs.

**Comment by xaionaro** [comment]
> Anyway, you removed stm32 hal conf header.

There are two problems with it.

1. It has these lines:
```
  * This software is licensed under terms that can be found in the LICENSE file
  * in the root directory of this software component.
  * If no LICENSE file comes with this software, it is provided AS-IS.
```
While the LICENSE file wasn't supplied. And I'm not sure what do they have in the modern day version of LICENSE file.

2. It hardcodes the project to STM32F4. It is fine if we make it extendable (so that owners of different MCUs have equal experience of using this library), but feels wrong to hardcode it to a specific family.

---

> Im planning to make next huge PR with refactored code and some changes. The headers and source will be in src/, but I will keep in mind requirement for other MCUs.

OK. Personally I don't have a strong opinion.

**Comment by mykola2312** [comment]
Arent all stm32 includes in stm32includes.h? Its matter of chaning few lines in CMake and stm32includes.h

**Comment by mykola2312** [comment]
> Im planning to make next huge PR with refactored code and some changes. The headers and source will be in src/, but I will keep in mind requirement for other MCUs.

Well it's no problem if headers are in Inc folder, CMake will anyway add them to include paths. I will think about backwards compatibility with v1

**Comment by mykola2312** [comment]
> Forgot to say one more thing: since you are the only active developer of this library in the whole world right now, it might be reasonable to adding you to collaborators, and may be move to a neutral org (so that it is not coupled with `xaionaro`) :)
> 
> About `inc`: I should've asked about `inc`-vs-`src` in the PR before merging it. The reason why I reverted it is just because I haven't found the reason why they were merged, while older projects already have configurations with `inc` (and I just didn't want to break them without reason). But if there is a good reason I'm happy to merge it back :)

Thanks for adding me to collaborators :)

**Comment by mykola2312** [comment]
I took another look at your changes and they are good no problem. I will sync my fork with your changes :)

**Comment by xaionaro** [comment]
> Arent all stm32 includes in stm32includes.h? Its matter of chaning few lines in CMake and stm32includes.h

Yeah, it is basically about improving `stm32includes.h` to include the MCU you use (the way other MCUs are included there) and improving the CMake configuration to allow to choose the MCU or something like that.

> Thanks for adding me to collaborators :)

I've moved the project to org `stm32-libs` and I've just sent out the invitation to join the org (so that you have write access to the repo if you accept the invitation).

**Comment by mykola2312** [comment]
Thanks. Now I should figure out how to change base repo of my fork. Do you by any chance know how to do that?

**Comment by xaionaro** [comment]
> Thanks. Now I should figure out how to change base repo of my fork. Do you by any chance know how to do that?

It should change automatically eventually. I understand it already did, right?

**Comment by mykola2312** [comment]
> > Thanks. Now I should figure out how to change base repo of my fork. Do you by any chance know how to do that?
> 
> It should change automatically eventually. I understand it already did, right?

Oh, you're right, I just needed to wait more. Thanks

**Pull request merged**
